src/framebuffer.rs

```diff
@@ -0,0 +1,707 @@
+//! Framebuffer.
+
+use core::{convert::Infallible, marker::PhantomData};
+
+use crate::{
+    draw_target::DrawTarget,
+    geometry::{OriginDimensions, Point, Size},
+    image::{ImagePixelGetter, ImageRaw},
+    iterator::raw::RawDataSlice,
+    pixelcolor::{
+        raw::{
+            BigEndian, ByteOrder, LittleEndian, RawData, RawU1, RawU16, RawU2, RawU24, RawU32,
+            RawU4, RawU8, ToBytes,
+        },
+        PixelColor,
+    },
+    Pixel,
+};
+
+/// Calculates the required buffer size.
+///
+/// This function is a workaround for current limitations in Rust const generics.
+/// It can be used to calculate the `N` parameter based on the size and color type of the framebuffer.
+pub const fn buffer_size<C: PixelColor>(width: usize, height: usize) -> usize {
+    (width * C::Raw::BITS_PER_PIXEL + 7) / 8 * height
+}
+
+/// A framebuffer.
+///
+/// # Examples
+///
+/// ```
+/// use embedded_graphics::{
+///     framebuffer,
+///     framebuffer::{Framebuffer, buffer_size},
+///     pixelcolor::{Rgb565, raw::LittleEndian},
+///     prelude::*,
+///     primitives::PrimitiveStyle,
+/// };
+///
+/// let mut fb = <framebuffer!(Rgb565, 320, 240)>::new();
+/// // or:
+/// let mut fb = Framebuffer::<Rgb565, _, LittleEndian, 320, 240, {buffer_size::<Rgb565>(320, 240)}>::new();
```

Comment:
Having to use a macro to determine the type isn't great, but IMO better than requiring the user to calculate 'N` manually.

Reply:
The non macro version is already much better than it used to be. Before we increased the MSRV the `buffer_size` function couldn't use a type parameter.